File: database/src/main/java/com/usharik/database/dao/DocumentDao.java
package com.usharik.database.dao;

import androidx.room.Dao;
import androidx.room.Query;

import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface DocumentDao {

    @Query("select count(*) from DOCUMENT")
    Single<Integer> getCount();

    @Query("select json from DOCUMENT where word = :word")
    Maybe<String> getJsonStringByWord(String word);

    @Query( """
            select *
            from DOCUMENT
            where declension_type != :declensionType
            order by random() limit 1
            """)
    Single<DocumentEntity> getRandomWordWithAnotherDeclensionType(String declensionType);
}

File: database/src/main/kotlin/com/usharik/database/dao/DocumentDao.kt
package com.usharik.database.dao

import androidx.room.Dao
import androidx.room.Query

@Dao
interface DocumentDao {
    @Query("select count(*) from DOCUMENT")
    suspend fun count(): Int

    @Query("select json from DOCUMENT where word = :word")
    suspend fun jsonForWord(word: String): String?

    @Query("""
        select * from DOCUMENT
        where declension_type != :declensionType
          and (:gender is null or gender = :gender)
        order by random() limit 1
    """)
    suspend fun randomWordWithAnotherDeclensionType(declensionType: String, gender: String?): DocumentEntity?
}
